### DISCUSSION TITLE
How to display two functions alternately.

### DISCUSSION THREAD

**Discussion by daktari77** [contrem/arduino-timer#69] [Q&A]
hello everyone, I hope you can help me. 
I am building a led clock which displays UTC time, local time, date and temperature / humidity. On the first line the UTC time timer.every (1000, funcA) is shown, on the second line the local time timer.every (1000, funcB) and on the third I would like to alternate the display of the date with temperature / humidity.

**Comment by skrobinson** [comment]
There are many ways to approach this. Here's mine. I hope others will share more ideas.

I assume you have `funcC` to show date and `funcD` to show temperature/humidity. Add a bool variable to track which function to call. Add a function to call `funcC` or `funcD` depending on the state of the bool. Change the tracking value to alternate which value to display on the third line.

```cpp
bool showDate = true;

bool funcE() {
    if (showDate) {
        showDate = false;
        return funcC();
    }
    showDate = true;
    return funcD();
}

timer.every (1000, funcE);
```

I have not tested this.

**Reply by philj404** [reply]
Here's another approach.  I like the @skrobinson approach better, but sometimes you might want something different:
```
#define NEXT_LINE2_UPDATE 1000

bool showDate(void)
{
   ...
   timer.in(NEXT_LINE2_UPDATE, showTimeAndHumidity);
}

bool showTimeAndHumidity(void)
{
   ...
   timer.in(NEXT_LINE2_UPDATE, showDate);
}

...
timer.in(NEXT_LINE2_UPDATE, showDate); // start either one of the tasks
...
```

This "state machine" approach can be used to build very interesting and complex behaviors -- particularly when the tasks add logic to choose what the next task should be.

**Reply by daktari77** [reply]
Many thanks philj404, I will test your suggestionin the next days.

**Reply by daktari77** [reply]
IT WORKS!!!!
thanks a lot.

**Reply by philj404** [reply]
I'm glad the "state machine" approach works for you.

The `timer.every(...);` example from @skrobinson can a bit simpler and easier to debug for simple alternating readouts.  It's obvious where the timer task is created and it never needs to change.

With the "state machine" example there are a few ways things can go wrong if you are not careful.  It may not be obvious what went wrong because the task with the problem finished and was deleted by the timer.  If you start to have problems like this, consider:

- If a task does not call `timer.in(...)` before it finishes, the next task will never run and readouts will freeze.  It may not be obvious which task did not make the call... or why.
- If there is no room for more tasks, the `timer.in(...)` call will fail.  Readouts will freeze for a different reason.
- If for some reason `timer.in(...)` gets called twice, the timer will have two of these tasks running.  Readouts may appear to stutter.  (For me my blinkLED() task seemed to toggle faster and faster...)

Good luck and have fun!

**Reply by skrobinson** [reply]
@philj404 Thank you. I had not thought of your approach.

The "state machine" has at least two advantages as I see it. It allows for non-uniform timing:

```cpp
#define NEXT_LINE2_SHORT_UPDATE 1000
#define NEXT_LINE2_LONG_UPDATE 2000

bool showDate(void)
{
   ...
   timer.in(NEXT_LINE2_SHORT_UPDATE, showTimeAndHumidity);
}

bool showTimeAndHumidity(void)
{
   ...
   timer.in(NEXT_LINE2_LONG_UPDATE, showDate);
}

timer.in(NEXT_LINE2_LONG_UPDATE, showDate);
```

And it is easy to form a chain of more than two display functions.

**Reply by daktari77** [reply]
exactly, in fact in this way I can display humidity and temperature for 10sec and the date for 3sec

**Comment by daktari77** [comment]
@skrobinson many thanks to you too, I tried both suggestions but the @philj404's idea required fewer changes to my code

**Comment by daktari77** [comment]
Guys, this is my wall clock project: https://github.com/daktari77/IU2GLR_Clock/tree/main/WALL_Clock_rev2.0b.01.

in case you want to take a look even just out of curiosity.
I am not a programmer, but I try to make do.

**Reply by philj404** [reply]
It looks interesting!  You may not be a programmer, but you have learned a _lot_ of concepts for this application.  I'm glad they are all coming together for you.

(I see you did NOT put your wifi/password on github.  Good for you!)

Which processor are you using?  I just learned of an ESP32 library which lets you connect to WiFi without hard-coding your network name or password.  It can make code more portable (maybe not important for you).  See: https://github.com/tzapu/WiFiManager if you are interested.

**Reply by daktari77** [reply]
Thanks. The board is an eps32 WROOM. I used 24 max72xx display modules and a DHT22.

This is my project 

![IMG_20221015_112555.jpg](https://user-images.githubusercontent.com/7127661/195979437-8d067240-a706-491b-bda4-36efdd3e7a5a.jpg)

![IMG_20221015_112605.jpg](https://user-images.githubusercontent.com/7127661/195979438-f220b47d-efeb-4963-b57d-c03fa0125cef.jpg)
